File: scripts/shared/inline-review-comments.mjs
#!/usr/bin/env node

import { githubRequest } from "./github-client.mjs";
import { formatFindingBody } from "./finding-format.mjs";

function parseRepository(repo) {
  const [owner, name] = String(repo || "").split("/");
  if (!owner || !name) return null;
  return { owner, name };
}

function normalizeText(value) {
  return String(value ?? "")
    .replace(/\r?\n+/g, " ")
    .replace(/\s{2,}/g, " ")
    .trim();
}

function normalizeFindingLine(value) {
  return Number.isInteger(value) && value > 0 ? value : 0;
}

export function isLineBoundFinding(finding) {
  if (!finding || typeof finding !== "object") return false;
  const file = normalizeText(finding.file);
  const line = normalizeFindingLine(finding.line);
  return file.length > 0 && line > 0;
}

export function buildInlineCommentBody({ reviewerLabel, finding }) {
  return formatFindingBody({ reviewerLabel, finding });
}

export async function publishInlineFindingComments({
  token,
  repo,
  prNumber,
  headSha,
  entries,
  labelsByReviewerId,
  request = githubRequest,
}) {
  const normalizedToken = normalizeText(token);
  const normalizedRepo = normalizeText(repo);
  const normalizedPrNumber = normalizeText(prNumber);
  const normalizedHeadSha = normalizeText(headSha);
  const normalizedEntries = Array.isArray(entries) ? entries : [];
  const labels =
    labelsByReviewerId instanceof Map
      ? labelsByReviewerId
      : new Map();

  if (!normalizedToken || !normalizedRepo || !normalizedPrNumber || !normalizedHeadSha) {
    return {
      attemptedCount: 0,
      postedCount: 0,
      failedCount: 0,
      postedEntries: [],
      failedEntries: [],
    };
  }

  const parsedRepo = parseRepository(normalizedRepo);
  if (!parsedRepo) {
    throw new Error("GITHUB_REPOSITORY must be owner/name");
  }

  const apiBase = `https://api.github.com/repos/${parsedRepo.owner}/${parsedRepo.name}`;
  const lineBoundEntries = normalizedEntries.filter((entry) => isLineBoundFinding(entry?.finding));
  const postedEntries = [];
  const failedEntries = [];

  for (const entry of lineBoundEntries) {
    const finding = entry.finding;
    const reviewer = normalizeText(entry.reviewer || "unknown");
    const reviewerLabel = normalizeText(labels.get(reviewer) || reviewer || "Unknown Reviewer");
    const body = buildInlineCommentBody({
      reviewerLabel,
      finding,
    });

    const path = normalizeText(finding.file);
    const line = normalizeFindingLine(finding.line);

    try {
      const created = await request({
        method: "POST",
        url: `${apiBase}/pulls/${normalizedPrNumber}/comments`,
        token: normalizedToken,
        body: {
          body,
          commit_id: normalizedHeadSha,
          path,
          line,
          side: "RIGHT",
        },
      });

      postedEntries.push({
        ...entry,
        comment_id: Number.isInteger(created?.id) && created.id > 0 ? created.id : null,
        comment_url: normalizeText(created?.html_url || ""),
      });
    } catch (error) {
      failedEntries.push({
        ...entry,
        error: normalizeText(error?.message || "unknown inline comment error"),
      });
    }
  }

  return {
    attemptedCount: lineBoundEntries.length,
    postedCount: postedEntries.length,
    failedCount: failedEntries.length,
    postedEntries,
    failedEntries,
  };
}

File: scripts/shared/inline_review_comments.py
from __future__ import annotations

from typing import Protocol

from .finding_format import format_finding_body
from .github_client import github_request
from .types import (
    InlineCommentEntry,
    JSONObject,
    PresentationEntry,
    PresentationFinding,
    PublishedInlineComments,
)


class RequestFunc(Protocol):
    def __call__(
        self,
        *,
        method: str,
        url: str,
        token: str,
        body: JSONObject | None = None,
    ) -> JSONObject | list[JSONObject] | str | int | float | bool | None: ...


def _parse_repository(repo: str) -> tuple[str, str] | None:
    parts = repo.strip().split("/")
    if len(parts) != 2 or not parts[0] or not parts[1]:
        return None
    return parts[0], parts[1]


def _normalize_text(value: str | int | float | bool | None) -> str:
    return " ".join(str("" if value is None else value).replace("\n", " ").split())


def _normalize_finding_line(value: int | None) -> int:
    return value if isinstance(value, int) and value > 0 else 0


def is_line_bound_finding(finding: PresentationFinding) -> bool:
    file_path = _normalize_text(finding.get("file"))
    line = _normalize_finding_line(finding.get("line"))
    return bool(file_path) and line > 0


def build_inline_comment_body(*, reviewer_label: str, finding: PresentationFinding) -> str:
    return format_finding_body(reviewer_label=reviewer_label, finding=finding)


def publish_inline_finding_comments(
    *,
    token: str,
    repo: str,
    pr_number: str,
    head_sha: str,
    entries: list[PresentationEntry],
    labels_by_reviewer_id: dict[str, str],
    request: RequestFunc = github_request,
) -> PublishedInlineComments:
    normalized_token = _normalize_text(token)
    normalized_repo = _normalize_text(repo)
    normalized_pr_number = _normalize_text(pr_number)
    normalized_head_sha = _normalize_text(head_sha)

    if not normalized_token or not normalized_repo or not normalized_pr_number or not normalized_head_sha:
        return PublishedInlineComments(
            attemptedCount=0,
            postedCount=0,
            failedCount=0,
            postedEntries=[],
            failedEntries=[],
        )

    parsed_repo = _parse_repository(normalized_repo)
    if not parsed_repo:
        raise ValueError("GITHUB_REPOSITORY must be owner/name")

    owner, name = parsed_repo
    api_base = f"https://api.github.com/repos/{owner}/{name}"
    line_bound_entries = [entry for entry in entries if is_line_bound_finding(entry["finding"])]

    posted_entries: list[InlineCommentEntry] = []
    failed_entries: list[InlineCommentEntry] = []

    for entry in line_bound_entries:
        finding = entry["finding"]
        reviewer = _normalize_text(entry.get("reviewer", "unknown"))
        reviewer_label = _normalize_text(labels_by_reviewer_id.get(reviewer, reviewer) or "Unknown Reviewer")
        body = build_inline_comment_body(reviewer_label=reviewer_label, finding=finding)

        path = _normalize_text(finding.get("file"))
        line = _normalize_finding_line(finding.get("line"))

        try:
            payload = request(
                method="POST",
                url=f"{api_base}/pulls/{normalized_pr_number}/comments",
                token=normalized_token,
                body={
                    "body": body,
                    "commit_id": normalized_head_sha,
                    "path": path,
                    "line": line,
                    "side": "RIGHT",
                },
            )

            comment_id: int | None = None
            comment_url = ""
            if isinstance(payload, dict):
                payload_id = payload.get("id")
                comment_id = payload_id if isinstance(payload_id, int) and payload_id > 0 else None
                comment_url_value = payload.get("html_url")
                comment_url = _normalize_text(comment_url_value if isinstance(comment_url_value, str) else "")

            posted_entries.append(
                InlineCommentEntry(
                    reviewer=entry["reviewer"],
                    status=entry["status"],
                    finding=entry["finding"],
                    comment_id=comment_id,
                    comment_url=comment_url,
                )
            )
        except Exception as error:
            failed_entries.append(
                InlineCommentEntry(
                    reviewer=entry["reviewer"],
                    status=entry["status"],
                    finding=entry["finding"],
                    error=_normalize_text(str(error)) or "unknown inline comment error",
                )
            )

    return PublishedInlineComments(
        attemptedCount=len(line_bound_entries),
        postedCount=len(posted_entries),
        failedCount=len(failed_entries),
        postedEntries=posted_entries,
        failedEntries=failed_entries,
    )
